CMakeLists.txt

```diff
@@ -1,71 +1,52 @@
-cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
-
-message(AUTHOR_WARNING "If the build fails, please try the autotools/qmake method.")
-
-if (POLICY CMP0074)
-    cmake_policy(SET CMP0074 NEW)
-endif()
-
-# TODO: fix the macOS bundle stuff
-if (POLICY CMP0080)
-    cmake_policy(SET CMP0080 OLD)
-endif()
+cmake_minimum_required(VERSION 3.16 FATAL_ERROR) # Policies <= CMP0097 default to NEW
 
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
-include(FunctionReadVersion)
-
-read_version("${CMAKE_CURRENT_SOURCE_DIR}/version.pri" VER_MAJOR VER_MINOR VER_BUGFIX VER_BUILD VER_STATUS)
-# message(STATUS "Project version is: ${VER_MAJOR}.${VER_MINOR}.${VER_BUGFIX}.${VER_BUILD} (${VER_STATUS})")
-
-project(qBittorrent VERSION ${VER_MAJOR}.${VER_MINOR}.${VER_BUGFIX}.${VER_BUILD})
-
-# check for invalid compiler version/CXX standard as early as possible
-include(FunctionQbtCXXCompilerAndModeCheck)
-qbt_minimum_cxx_mode_check(14)
-message(STATUS "Building in C++${CMAKE_CXX_STANDARD} mode.\n"
-            "Make sure libtorrent was built with the same C++ mode for ABI compatibility.")
-
-set(PROJECT_VERSION "${VER_MAJOR}.${VER_MINOR}.${VER_BUGFIX}")
-
-if (NOT VER_BUILD EQUAL 0)
-    set(PROJECT_VERSION "${PROJECT_VERSION}.${VER_BUILD}")
-endif()
 
-set(PROJECT_VERSION "${PROJECT_VERSION}${VER_STATUS}")
-
-add_definitions(-DQBT_VERSION_MAJOR=${VER_MAJOR})
-add_definitions(-DQBT_VERSION_MINOR=${VER_MINOR})
-add_definitions(-DQBT_VERSION_BUGFIX=${VER_BUGFIX})
-add_definitions(-DQBT_VERSION_BUILD=${VER_BUILD})
-
-add_definitions(-DQBT_VERSION="v${PROJECT_VERSION}")
-add_definitions(-DQBT_VERSION_2="${PROJECT_VERSION}")
-
-include(GNUInstallDirs)
-include(FeatureSummary)
+project(qBittorrent
+    VERSION 4.3.0.0
+    DESCRIPTION "The qBittorrent BitTorrent client"
+    HOMEPAGE_URL "https://www.qbittorrent.org/"
+    LANGUAGES CXX
+)
 
+# use CONFIG mode first in find_package
+set(CMAKE_FIND_PACKAGE_PREFER_CONFIG ON)
 # version requirements
 set(requiredBoostVersion 1.40)
 set(requiredQtVersion 5.9.0)
-set(requiredOpensslVersion 1.0)
-set(requiredLibtorrentVersion 1.2)
+set(requiredOpenSSLVersion 1.0)
+set(requiredLibtorrentVersion 1.2.0)
+set(requiredZlibVersion 1.2.5.2)
 
-if (WIN32)
-    include(winconf)
+# features (some are platform-specific)
+include(FeatureSummary)
+include(FeatureOptionsSetup)
+feature_option(BUILD_SHARED_LIBS "Build internal libraries SHARED instead of STATIC (for experimental development purposes)" OFF)
```

Comment:
I don't remember it being present the last time I reviewed it (or I didn't notice it because of other issues).
It is very questionable (as well as current splitting into several libraries), so I would remove it before it becomes a new stumbling block here.

Reply:
@glassez 

`BUILD_SHARED_LIBS` is a built-in CMake construct. It makes every "naked" `add_library()` call (i.ie., one that **does not** specify `STATIC` or `SHARED`) act as `STATIC` or `SHARED`, depending on whether it is `OFF` (default) or `ON`, respectively. Wrapping it in inside `feature_option()` is just a convenience to have its description printed out - it makes sense to me for it to be printed, since it provides such important functionality and is already there.

When it's off (the default) it doesn't change anything that we were already doing.

Declaring libraries with "naked" `add_library()` is good for flexibility, because it gives the builder the choice via the simple meachanism of passing `-DBUILD_SHARED_LIBS=ON/OFF`. Currently, qBittorrent does not build at all with `-DBUILD_SHARED_LIBS=ON` (hence the note about "experimental development purposes"), but this may change in the future, unless we decide that there is no `add_library()` call in our codebase that would ever make sense to offer such flexibility.

Docs: https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html

Reply:
Currently qBittorrent isn't supposed to be a set of several libraries.
The fact that CMake project divides it into several (static) libraries can only be considered as a workaround that allows CMake to follow the file structure of the project's source code (perhaps due to some drawbacks that prevent it from being done in another way).
You should not change this here, especially since the build as a set of shared libraries is not possible currently (at least on Windows, it requires changes to the source code).

Reply:
@glassez 

> Currently qBittorrent isn't supposed to be a set of several libraries.
The fact that CMake project divides it into several (static) libraries can only be considered as a workaround that allows CMake to follow the file structure of the project's source code (perhaps due to some drawbacks that prevent it from being done in another way).

When shared libraries are out of the question, the Professional CMake book still recommends building projects as a set of static libraries, as opposed to using plain `OBJECT` libraries, unless there is a special reason that requires its semantics.

> You should not change this here,

I did not change anything - like I said - `BUILD_SHARED_LIBS` is something that always exists, even if you don't do anything, and it's `OFF` by default. Like previously, the project is till built as a set of static libraries by default.

> especially since the build as a set of shared libraries is not possible currently (at least on Windows, it requires changes to the source code).

Yes, I know such a build is currently not possible, hence the "(for experimental development purposes)" part of the description. No one would normally turn the feature on.

The question becomes: do you think such a build with shared libraries will never be possible, or that it doesn't make sense to indicate that one day it might be? If so, it is a good idea to remove this. Otherwise, we can keep it in, it might encourage someone to investigate and submit patches to make it happen.

Reply:
@glassez 

Well, to be honest I don't really care enough about this change to keep pushing for it right now. Its potential benefits for the near and medium term future are unclear, and if we do need such a change for the far future, its quite trivial anyway. So I reverted the internal libraries to be explicitly `STATIC`, like previously, and removed the `BUILD_SHARED_LIBS` description from the output.

Reply:
>So I reverted the internal libraries to be explicitly STATIC, like previously, and removed the BUILD_SHARED_LIBS description from the output.

:+1: